ion-schema/src/isl/isl_constraint.rs

```diff
@@ -148,54 +209,317 @@ impl IslConstraint {
+        )
+    }
+
+    /// Creates an [IslConstraint::TimestampOffset] using the offset list specified in it
+    pub fn timestamp_offset(offsets: Vec<TimestampOffset>) -> IslConstraint {
+        IslConstraint::new(
+            IonSchemaLanguageVersion::V2_0,
+            IslConstraintImpl::TimestampOffset(IslTimestampOffsetConstraint::new(offsets)),
+        )
+    }
+
+    /// Creates an [IslConstraint::Utf8ByteLength] using the range specified in it
+    pub fn utf8_byte_length(length: NonNegativeIntegerRange) -> IslConstraint {
+        IslConstraint::new(
+            IonSchemaLanguageVersion::V2_0,
+            IslConstraintImpl::Utf8ByteLength(Range::NonNegativeInteger(length)),
+        )
+    }
+
+    /// Creates an [IslConstraint::Element] using the [IslTypeRef] referenced inside it
+    pub fn element(isl_type: IslTypeRef) -> IslConstraint {
+        // add support for `distinct::` elements
+        todo!()
+    }
+
+    /// Creates an [IslConstraint::Annotations] using [str]s and [Element]s specified inside it
+    pub fn annotations<'a, A: IntoIterator<Item = &'a str>, B: IntoIterator<Item = Element>>(
+        annotations_modifiers: A,
+        annotations: B,
+    ) -> IslConstraint {
+        todo!()
+    }
+
+    /// Creates a [IslConstraint::ValidValues] using the [Element]s specified inside it
+    pub fn valid_values_with_values(values: Vec<Element>) -> IonSchemaResult<IslConstraint> {
+        let valid_values: IonSchemaResult<Vec<ValidValue>> =
+            values.iter().map(|e| e.try_into()).collect();
+        Ok(IslConstraint::new(
+            IonSchemaLanguageVersion::V2_0,
+            IslConstraintImpl::ValidValues(IslValidValuesConstraint {
+                valid_values: valid_values?,
+            }),
         ))
     }
 
+    /// Creates a [IslConstraint::ValidValues] using the [Range] specified inside it
+    pub fn valid_values_with_range(range: Range) -> IslConstraint {
+        IslConstraint::new(
+            IonSchemaLanguageVersion::V2_0,
+            IslConstraintImpl::ValidValues(IslValidValuesConstraint {
+                valid_values: vec![ValidValue::Range(range)],
+            }),
+        )
+    }
+
+    /// Creates an [IslConstraint::Regex] using the expression and flags (case_insensitive, multi_line)
+    pub fn regex(case_insensitive: bool, multi_line: bool, expression: String) -> IslConstraint {
+        todo!()
+    }
+}
+
+/// Represents schema constraints [IslConstraint] which stores IslTypeRef
+#[derive(Debug, Clone, PartialEq)]
+pub struct IslConstraint {
+    pub(crate) version: IonSchemaLanguageVersion,
+    pub(crate) constraint: IslConstraintImpl,
+}
+
+impl IslConstraint {
+    pub(crate) fn new(version: IonSchemaLanguageVersion, constraint: IslConstraintImpl) -> Self {
+        Self {
+            constraint,
+            version,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum IslConstraintImpl {
```

Comment:
I wish there were a more descriptive name for this, but I'm having difficulty thinking of one. Fortunately, since this is a private type we can rename it later if inspiration strikes.